Nodejs/Product/Nodejs/Workspace/ContextMenuProvider.cs

```diff
@@ -0,0 +1,238 @@
+                using (var npmWorker = new NpmWorker(npmController))
+                using (var manager = new NpmPackageInstallWindow(npmController, npmWorker))
+                {
+                    manager.ShowModal();
+                }
+            }
+
+            private async void ExecNpmUpdate(WorkspaceVisualNodeBase node)
+            {
+                using (var npmController = this.CreateController(node.Workspace))
+                using (var commander = npmController.CreateNpmCommander())
+                {
+                    await commander.UpdatePackagesAsync();
+                }
+            }
+
+            private async void ExecDynammic(WorkspaceVisualNodeBase node, uint nCmdID)
+            {
+                // Unfortunately the NpmController (and NpmCommander), used for the install, update commands
+                // doesn't support running arbitrary scripts. And changing that is outside
+                // the scope of these changes.
+
+                var index = nCmdID - PkgCmdId.cmdidWorkSpaceNpmDynamicScript;
+                var packageJson = PackageJsonFactory.Create(((IFileNode)node).FullPath);
+
+                Debug.Assert(packageJson != null, "Failed to create package.json");
+
+                var scripts = packageJson.Scripts;
+                if (index < scripts.Length)
+                {
+                    var npmPath = NpmHelpers.GetPathToNpm();
+
+                    await NpmWorker.ExecuteNpmCommandAsync(
+                        npmPath,
+                        executionDirectory: node.Workspace.Location,
+                        arguments: new[] { "run-script", scripts[index].CommandName },
+                        visible: true); // show the CMD window
+                }
+            }
+
+            public bool QueryStatus(List<WorkspaceVisualNodeBase> selection, Guid pguidCmdGroup, uint nCmdID, ref uint cmdf, ref string customTitle)
+            {
+                if (selection.Count != 1 || !EnsurePackageJson(selection.First()))
+                {
+                    return false;
+                }
+
+                if (pguidCmdGroup == Guids.NodeToolsWorkspaceCmdSet)
+                {
+                    switch (nCmdID)
+                    {
+                        // we can always install missing, new, and update npm packages
+                        case PkgCmdId.cmdidWorkSpaceNpmInstallMissing:
+                        case PkgCmdId.cmdidWorkSpaceNpmInstallNew:
+                        case PkgCmdId.cmdidWorkSpaceNpmUpdate:
+                            cmdf = (uint)(OLECMDF.OLECMDF_SUPPORTED | OLECMDF.OLECMDF_ENABLED);
+                            return true;
+                    }
+
+                    // Each id we return true for, this method is called with id + 1
+                    // this way we can add each script to the context menu
+                    if (nCmdID >= PkgCmdId.cmdidWorkSpaceNpmDynamicScript && nCmdID < PkgCmdId.cmdidWorkSpaceNpmDynamicScriptMax)
```

Comment:
>cmdidWorkSpaceNpmDynamicScriptMax [](start = 95, length = 33)

Generally, I would expect "Max" to be included in the range, so this might be a misnomer.

Reply:
changed